serve/mlc_serve/engine/staging_engine_worker.py

```diff
@@ -194,24 +197,25 @@ def step(self) -> GenerationLoopWorkerOutput:
 
         self.stopped_requests.clear()
 
-        for state in self.cancelled_requests:
-            err = None
-            if state.validation_err:
-                err = state.validation_err
+        with self.queue_lock:
```

Comment:
Why do we need this lock?

Reply:
I got an impression that I observed some data race during debugging when we had requests which should not be in the `self.cancelled_requests`. Probably because we add requests into self.canceled_requests in [add()](https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/engine/staging_engine_worker.py#L118C24-L118C24) and remove in [step()](https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/engine/staging_engine_worker.py#L214) that I assume can be called concurrently from different threads. I guarded place in `step()`, place in `add()` was guarded before

Reply:
yeah that makes sense.